packages/frontend/src/components/ChooseAppAndEventSubstep/index.tsx

```diff
@@ -93,24 +90,63 @@ function ChooseAppAndEventSubstep(
   const appOptions = useMemo(
     () =>
       apps
-        // Filter away stuff hidden behind feature flags
         ?.filter((app) => {
+          //
+          // ** EDGE CASE **
+          //
+          // We edge case since a generic implementation adds too much
+          // complexity; we'll move to generic if there's another use case for
+          // such hiding.
+          //
+          // If everyone forgets about this, it's OK because the next guy to
+          // add a new toolbox action will get confused why toolbox is missing
+          // ... and find this.
+          //
+          if (
+            app.key === TOOLBOX_APP_KEY &&
+            !isIfThenSelectable(allEditorSteps, step, launchDarkly?.flags)
+          ) {
+            return false
+          }
+
+          // Filter away stuff hidden behind feature flags
           if (!launchDarkly.flags || !app?.key) {
             return true
           }
           const launchDarklyKey = ['app', app.key].join('_')
           return launchDarkly.flags[launchDarklyKey] ?? true
         })
-        .map((app) => optionGenerator(app)) ?? [],
-    [apps, launchDarkly.flags],
+        ?.map((app) => optionGenerator(app)) ?? [],
+    [apps, step.position, allEditorSteps, launchDarkly.flags],
   )
+
   const actionsOrTriggers: Array<ITrigger | IAction> =
     (isTrigger ? app?.triggers : app?.actions) || []
   const actionOrTriggerOptions = useMemo(
     () =>
       actionsOrTriggers
-        // Filter away stuff hidden behind feature flags
+
         .filter((actionOrTrigger) => {
+          //
+          // ** EDGE CASE AGAIN **
+          //
+          // Hello, the if-then edge case demon here again!
+          //
+          // To ensure if-then is always the last step, we also need to guard
+          // against users selecting toolbox first, then adding steps after the
+          // toolbox step, then selecting If-Then in the toolbox step.
+          //
+          // Luckily, we can remove the top app-hiding edge case once we
+          // implement for-each, and toolbox will have multiple actions.
+          //
+          if (
+            app?.key === TOOLBOX_APP_KEY &&
+            !isIfThenSelectable(allEditorSteps, step, launchDarkly?.flags)
```

Comment:
check for `actionOrTrigger.key` as well, so it doesnt affect other toolbox actions in future

Reply:
these 3 checks can actually be done in this file,
1. is in branch? --> check depth from DepthContext
2. is last step --> check isLastStep prop
3. LaunchDarkly flag

Reply:
Haha actually I was passing in all these to siam using `DepthContext` (since it felt a bit dirty to use another context)

But seeing your comment - if we're edge casing anyway, might as well make the edge case simple. Agreed on suggestion